src/components/05_pages/Permissions/Permissions.js

```diff
@@ -1,17 +1,156 @@
+import makeCancelable from 'makecancelable';
+import { Markup } from 'interweave';
 
-const styles = {
-  title: css`
-    text-decoration: underline;
-  `,
-};
+import Loading from '../../Helpers/Loading';
+import { Table, TableBody, TableHeaderSimple } from '../../UI';
+
+const Permissions = class Permissions extends Component {
+  state = {
+    loaded: false,
+  };
+  componentDidMount() {
+    this.cancelFetch = makeCancelable(
+      Promise.all([
+        fetch(
+          `${
+            process.env.REACT_APP_DRUPAL_BASE_URL
+          }/admin-api/permissions?_format=json`,
+        ).then(res => res.json()),
+        fetch(
+          `${
+            process.env.REACT_APP_DRUPAL_BASE_URL
+          }/jsonapi/user_role/user_role`,
+        ).then(res => res.json()),
+      ])
+        .then(([permissions, { data: roles }]) =>
+          this.setState({
+            rawPermissions: Object.keys(permissions).map(
+              key => permissions[key],
+            ),
+            renderablePermissions: Object.keys(permissions).map(
+              key => permissions[key],
+            ),
+            changedRoles: [],
+            // Move admin roles to the right.
+            roles: roles.sort((a, b) => {
+              if (a.attributes.is_admin && b.attributes.is_admin) {
+                return a.attributes.id - b.attributes.id;
+              } else if (a.attributes.is_admin) {
+                return 1;
+              } else if (b.attributes.is_admin) {
+                return -1;
+              }
+              return a.attributes.id - b.attributes.id;
+            }),
+            loaded: true,
+          }),
+        )
+        .catch(err => this.setState({ loaded: false, err })),
+    );
+  }
+  componentWillUnmount() {
+    this.cancelFetch();
+  }
+  onPermissionCheck = (roleName, permission) => {
+    this.setState(prevState => ({
+      changedRoles: [...new Set(prevState.changedRoles).add(roleName).values()],
+      roles: this.togglePermission(permission, roleName, prevState.roles),
+    }));
+  };
+
+  togglePermission = (permission, roleName, roles) => {
+    const roleIndex = roles.map(role => role.attributes.id).indexOf(roleName);
+    const role = roles[roleIndex];
+    const index = role.attributes.permissions.indexOf(permission);
+    if (index !== -1) {
+      role.attributes.permissions.splice(index, 1);
+    } else {
+      role.attributes.permissions.push(permission);
+    }
+    roles[roleIndex] = role;
+    return roles;
+  };
 
-const Permissions = () => (
-  <div>
-    <h1 className={styles.title}>Permissions</h1>
-    <p>This will be the permissions page.</p>
-  </div>
-);
+  groupPermissions = permissions =>
+    Object.entries(
+      permissions.reduce((acc, cur) => {
+        acc[cur.provider] = acc[cur.provider] || [];
+        acc[cur.provider].push(cur);
+        return acc;
+      }, {}),
+    );
+
+  createTableRows = (groupedPermissions, roles) =>
+    [].concat(
```

Comment:
Why `[].concat()` instead of array spreads? It seems you prefer spreads elsewhere.

Reply:
For clarity, I'm not after any change here, I'm genuinely curious,

Reply:
@e0ipso It's a bit subtle, but this is the easiest way to flatten the two-dimensional array that's created from constructing the table rows.